File: bottomsheetmenu/src/main/java/com/strekha/bottomsheetmenu/BottomMenu.java
package com.strekha.bottomsheetmenu;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.ColorStateList;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.internal.view.SupportMenuItem;
import android.support.v7.content.res.AppCompatResources;
import android.support.v7.view.menu.MenuBuilder;
import android.support.v7.view.menu.MenuItemImpl;
import android.util.TypedValue;
import android.view.MenuItem;

import java.util.ArrayList;
import java.util.List;

@SuppressLint("RestrictedApi")
class BottomMenu extends MenuBuilder {

    private static final int[] EMPTY_STATE_SET = {};

    @Nullable
    private ColorStateList iconColorStateList;

    BottomMenu(@NonNull Context context, @NonNull TintWrapper tintWrapper) {
        super(context);
        iconColorStateList = tintWrapper == TintWrapper.DEFAULT_TINT
                ? createDefaultTint(android.R.attr.textColorSecondary)
                : tintWrapper.tint;
    }

    @Override
    protected MenuItem addInternal(int group, int id, int categoryOrder, CharSequence title) {
        SupportMenuItem item = ((SupportMenuItem) super.addInternal(group, id, categoryOrder, title));
        item.setIconTintList(iconColorStateList);
        return item;
    }

    @NonNull
    List<SupportMenuItem> getItems() {
        List<MenuItemImpl> visibleItems = getVisibleItems();
        List<SupportMenuItem> items = new ArrayList<>(visibleItems.size());
        items.addAll(visibleItems);
        return items;
    }

    private ColorStateList createDefaultTint(int baseColorThemeAttr) {
        final TypedValue value = new TypedValue();
        if (!getContext().getTheme().resolveAttribute(baseColorThemeAttr, value, true)) {
            return null;
        }
        ColorStateList baseColor = AppCompatResources.getColorStateList(
                getContext(),
                value.resourceId
        );
        int defaultColor = baseColor.getDefaultColor();
        return new ColorStateList(
                new int[][]{EMPTY_STATE_SET},
                new int[]{defaultColor}
        );
    }

    static class TintWrapper {

        static final TintWrapper DEFAULT_TINT = new TintWrapper(null);

        @Nullable
        final ColorStateList tint;

        TintWrapper(@Nullable ColorStateList tint) {
            this.tint = tint;
        }
    }
}

File: bottomsheetmenu/src/main/java/com/strekha/bottomsheetmenu/BottomMenu.kt
package com.strekha.bottomsheetmenu

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.ColorStateList
import android.support.v4.internal.view.SupportMenuItem
import android.support.v7.content.res.AppCompatResources
import android.support.v7.view.menu.MenuBuilder
import android.support.v7.view.menu.MenuItemImpl
import android.util.TypedValue
import android.view.MenuItem

import java.util.ArrayList

@SuppressLint("RestrictedApi")
internal class BottomMenu(context: Context, tintWrapper: TintWrapper) : MenuBuilder(context) {

    private val iconColorStateList: ColorStateList?

    val items: List<SupportMenuItem>
        get() {
            val visibleItems = visibleItems
            val items = ArrayList<SupportMenuItem>(visibleItems.size)
            items.addAll(visibleItems)
            return items
        }

    init {
        iconColorStateList = if (tintWrapper === TintWrapper.DEFAULT_TINT)
            createDefaultTint(android.R.attr.textColorSecondary)
        else
            tintWrapper.tint
    }

    override fun addInternal(group: Int, id: Int, categoryOrder: Int, title: CharSequence): MenuItem {
        val item = super.addInternal(group, id, categoryOrder, title) as SupportMenuItem
        item.iconTintList = iconColorStateList
        return item
    }

    private fun createDefaultTint(baseColorThemeAttr: Int): ColorStateList? {
        val value = TypedValue()
        if (!context.theme.resolveAttribute(baseColorThemeAttr, value, true)) {
            return null
        }
        val baseColor = AppCompatResources.getColorStateList(
                context,
                value.resourceId
        )
        return ColorStateList(
                arrayOf(EMPTY_STATE_SET),
                intArrayOf(baseColor.defaultColor)
        )
    }

    internal class TintWrapper(val tint: ColorStateList?) {

        companion object {
            val DEFAULT_TINT = TintWrapper(null)
        }
    }

    companion object {
        private val EMPTY_STATE_SET = intArrayOf()
    }
}
